cql3/selection/selection.cc

```diff
@@ -386,17 +388,61 @@ std::unique_ptr<result_set> result_set_builder::build() {
     return std::move(_result_set);
 }
 
-result_set_builder::restrictions_filter::restrictions_filter(::shared_ptr<restrictions::statement_restrictions> restrictions,
+namespace {
+
+using namespace expr;
+
+/// A visitor that copies an expression, but replaces token operator and unselected-column operators with the
+/// TRUE literal.
+struct drop_token_and_unselected_columns {
+    const selection& sel;
+
+    expression operator()(const conjunction& c) const {
+        auto r = c.children | std::views::transform([this] (const expression& e) { return std::visit(*this, e); });
+        return conjunction{std::vector<expression>(r.begin(), r.end())};
+    }
+
+    expression operator()(const binary_operator& binop) const {
+        return seastar::visit(binop.lhs,
+                [] (const token&) { return expression(true); },
```

Comment:
What about something like `token(x) != :marker`? It still has to be filtered. Or maybe it's translated to two intervals? Or not supported at all?

Reply:
Maybe I'm missing why we want to drop tokens and unselected columns from the expression.

Reply:
> What about something like `token(x) != :marker`?

Forbidden.
`InvalidRequest: Error from server: code=2200 [Invalid query] message="Unsupported "!=" relation: token(p) != 2"` 

> Maybe I'm missing why we want to drop tokens and unselected columns from the expression.

Tokens are used to prepare the fetch command; those subexpressions can never be false on the fetched rows.  Unselected columns will not be fetched, so we cannot get their values to evaluate the filter expression.  See old code:

https://sourcegraph.com/github.com/scylladb/scylla/-/blob/cql3/selection/selection.cc#L430

Reply:
Can we have an expression on an unselected column? I can think of two cases:

 1. WHERE clause, but constraint is on a pk/ck column so moved to slice and not added to selection
 2. WHERE clause, constraint on regular column, added to selection even if not used.

Is this correct? Either way, please add a comment explaining why we drop unselected columns and why it's safe.

Reply:
This is roughly correct (though note that some pk restrictions still require filtering, and it makes a difference if there are unrestricted primary columns). I'll add a comment.